File: lib/models/assistant_state.dart
import 'package:vixrex/config/vitrin_alanlari.g.dart';
import 'package:vixrex/models/chat_message.dart';
import 'package:vixrex/services/vixrex_guidance_service.dart';
import 'package:vixrex/services/vixrex_profile_snapshot.dart';

/// VixRex'in "sırada ne var" kararının tek biçimi.
///
/// Faz E (Tek Asistan planı): bugün aynı soruyu üç ayrı tip yanıtlıyor —
/// `VixRexProfileSnapshot` (doluluk), `VixRexRecommendation` (öneri),
/// `VixRexQualityReport` (kalite puanı). Bu tip onları BİRLEŞTİRMEZ; bu
/// fazın kuralı "karar değişmez, kararın biçimi tekilleşir" — asıl birleşme
/// Faz F'de. `AssistantState.fromSnapshot` yalnız mevcut ikisinin
/// (`snapshot` + `VixRexGuidanceService.recommendationFor`) sonucunu tek bir
/// nesneye kopyalar; kendi kararını üretmez.
///
/// [mesajAnahtari] not: Faz B'nin `shared/vixrex_mesajlar.json` kataloğu şu
/// an yalnız sohbet-niyeti yanıtlarını taşıyor (bkz. `chatbot_config.dart`),
/// rehberlik önerilerinin başlık/açıklamasını değil. Bu alan bugün
/// `VixRexRecommendation.id` değerini taşır — zaten kararlı, benzersiz bir
/// anahtar. Bu id'leri Next.js'in de okuyabileceği JSON'a taşımak (metnin
/// TEK yerde yaşaması) Faz G3'ün işi; burada yapılmadı, kapsam dışı
/// bırakıldı.
class AssistantState {
  const AssistantState({
    required this.asama,
    required this.sonrakiEksikAlan,
    required this.sonrakiAdim,
    required this.mesajAnahtari,
    required this.eylemler,
    required this.doluluk,
    required this.doluAlan,
    required this.toplamAlan,
  });

  /// VixRex'in şu anki rehberlik aşaması (kurulum / yayın / paylaşım /
  /// geliştirme).
  final VixRexJourneyPhase asama;

  /// Sıradaki eksik ZORUNLU alan — şemadan, sırası da şemadan. `null` ise
  /// zorunlu alan eksiği yok (yasal onay veya yayın aşamasında olabilir).
  final VitrinAlani? sonrakiEksikAlan;

  /// Akış aşaması dâhil sıradaki adım (`legal`/`publish`/`share` alan
  /// değildir, zorunlu-alan kümesinin ötesindeki akış duraklarıdır).
  final VixRexNextStep sonrakiAdim;

  /// Öneri metninin anahtarı — bkz. sınıf başı not.
  final String mesajAnahtari;

  /// Sıralı eylem listesi, ilki birincil. Bugünkü `VixRexRecommendation`
  /// tek eylem taşıdığı için bu liste her zaman tek elemanlı; çoklu eylem
  /// desteği eklendiğinde imza değişmeden genişler.
  final List<AssistantAction> eylemler;

  /// 0-100 arası doluluk yüzdesi — zorunlu alan + yasal onay kümesi
  /// üzerinden.
  final int doluluk;

  /// Tamamlanan zorunlu adım sayısı.
  final int doluAlan;

  /// Toplam zorunlu adım sayısı.
  final int toplamAlan;

  /// Mevcut `VixRexProfileSnapshot` + `VixRexGuidanceService` çıktısını tek
  /// nesneye kopyalar. `snapshot` null olabilir (henüz hiç veri yok) —
  /// `VixRexGuidanceService.recommendationFor` zaten bu durumu biliyor,
  /// burada tekrar dallanmaz.
  factory AssistantState.fromSnapshot(
    VixRexProfileSnapshot? snapshot, {
    required bool hasShared,
  }) {
    final rec = VixRexGuidanceService.recommendationFor(
      snapshot: snapshot,
      hasShared: hasShared,
    );
    final toplamAlan =
        snapshot?.totalRequiredStepCount ?? (zorunluAlanlar.length + 1);
    final doluAlan = snapshot?.completedRequiredStepCount ?? 0;

    return AssistantState(
      asama:
          snapshot?.journeyPhase(hasShared: hasShared) ??
          VixRexJourneyPhase.setup,
      sonrakiEksikAlan: snapshot?.sonrakiEksikZorunluAlan,
      sonrakiAdim: snapshot?.nextMissingField ?? VixRexNextStep.name,
      mesajAnahtari: rec.id,
      eylemler: [
        AssistantAction(
          action: rec.action,
          label: rec.buttonLabel,
          primary: true,
        ),
      ],
      doluluk:
          toplamAlan == 0
              ? 0
              : ((doluAlan / toplamAlan) * 100).round().clamp(0, 100),
      doluAlan: doluAlan,
      toplamAlan: toplamAlan,
    );
  }
}

/// [AssistantState.eylemler] öğesi — `VixRexAction`'ı ekranın çizeceği
/// etiketle birlikte taşır.
class AssistantAction {
  const AssistantAction({
    required this.action,
    required this.label,
    this.primary = false,
  });

  final VixRexAction action;
  final String label;
  final bool primary;
}

File: public_web/src/lib/assistantState.ts
// Vixrex'in "sırada ne var" kararının Next.js tarafındaki biçimi.
//
// Faz E (Flutter, lib/models/assistant_state.dart) ile Faz G3 arasındaki
// boşluğu kapatır — PLAN.md Faz G3: "hazirlikRaporu çıktısı AssistantState'in
// TS ikizine map edilir (lib/assistantState.ts) — aynı alan adları, aynı
// anlam." Bu dosya o eşlemeyi kurar.
//
// DÜRÜSTLÜK NOTU: Dart'ın AssistantState'i Flutter'ın SOHBET AKIŞI
// (kurulum → yayın → paylaşım) etrafında kurulu — VixRexNextStep enum'u
// "name/whatsapp/address/legal/publish/share" gibi akış duraklarını taşır.
// Next.js'in sahip paneli farklı bir bağlamda çalışır: buraya yalnız
// ZATEN VAR olan (taslağı oluşturulmuş) bir vitrin için gelinir — "önce
// yayınla" gibi bir akış durağı burada anlamsız. Bu yüzden alan adları
// birebir aynı DEĞİL; anlamca karşılık gelen kısım (doluluk, sıradaki eksik
// alan, birincil eylem) aktarıldı, Flutter'a özgü akış kavramları
// (VixRexNextStep, publish/share fazları) zorla taşınmadı.
import {
  hazirlikRaporu,
  sonrakiRehberAlanlar,
  type EksikAlan,
  type HazirlikRaporu,
} from "./vitrinReadiness";

/** Dart'ın `VixRexJourneyPhase`sinin Next.js'teki karşılığı — yalnız iki
 * hâl var çünkü sahip paneline yalnız var olan bir taslak için gelinir. */
export type AssistantJourneyPhase = "kurulum" | "gelistirme";

/** [AssistantState.eylemler] öğesi. Dart'ın `VixRexAction` enum'u (ekran
 * navigasyonu) burada karşılığı yok — Next.js'te "eylem" bir alana
 * odaklanmak/kaydırmaktır, `hedefAlan` bu yüzden bir enum değil şema
 * anahtarı taşır. */
export interface AssistantAction {
  hedefAlan: string | null;
  label: string;
  primary: boolean;
}

export interface AssistantState {
  asama: AssistantJourneyPhase;
  /** Sıradaki eksik alan — şemadan, sırası da şemadan. `null` ise eksik yok. */
  sonrakiEksikAlan: EksikAlan | null;
  /** Öneri metninin anahtarı. Faz B kataloğu (shared/vixrex_mesajlar.json)
   * şu an yalnız sohbet-niyet yanıtlarını taşıyor, rehberlik önerisi
   * metnini değil (Dart tarafındaki aynı not — bkz.
   * lib/services/vixrex_guidance_service.dart). Bu alan bugün
   * `alan_<anahtar>` biçiminde kararlı bir kimlik taşır; gerçek metin
   * `hazirlikRaporu().sonrakiAdim`'den (zaten üretilmiş cümle) gelir. */
  mesajAnahtari: string;
  /** Sıralı eylem listesi, ilki birincil. */
  eylemler: AssistantAction[];
  doluluk: number;
  doluAlan: number;
  toplamAlan: number;
}

/** Zaten hesaplanmış bir `HazirlikRaporu`dan `AssistantState` üretir —
 * kendi kararını üretmez, yalnız biçimini tekilleştirir (Faz E'nin Dart
 * tarafındaki kuralıyla aynı). */
export function assistantStateFromRapor(rapor: HazirlikRaporu): AssistantState {
  const ilk = rapor.eksikler[0] ?? null;

  return {
    asama: rapor.temelTamam ? "gelistirme" : "kurulum",
    sonrakiEksikAlan: ilk,
    mesajAnahtari: ilk ? `alan_${ilk.anahtar}` : "tum_alanlar_tamam",
    eylemler: ilk
      ? [{ hedefAlan: ilk.anahtar, label: `${ilk.etiket} ekle`, primary: true }]
      : [],
    doluluk: rapor.yuzde,
    doluAlan: rapor.doluSayisi,
    toplamAlan: rapor.toplamSayisi,
  };
}

/** `draftData`'dan doğrudan `AssistantState` üretir — `hazirlikRaporu` +
 * `assistantStateFromRapor`'u art arda çağırmanın kısayolu. */
export function assistantStateFromDraft(
  draftData: Record<string, unknown>,
  atlanmislar: ReadonlySet<string> = new Set(),
): AssistantState {
  return assistantStateFromRapor(hazirlikRaporu(draftData, atlanmislar));
}

/** "Sırada" listesi (Faz G3, `UpNextList`) için — sonraki [adet] eksik
 * alanı birincil eylem biçiminde döner. */
export function upcomingActions(
  draftData: Record<string, unknown>,
  suankiAnahtar: string | null,
  atlanmislar: ReadonlySet<string>,
  adet: number = 3,
): AssistantAction[] {
  return sonrakiRehberAlanlar(draftData, suankiAnahtar, atlanmislar, adet).map(
    (alan, index) => ({
      hedefAlan: alan.anahtar,
      label: alan.etiket,
      primary: index === 0,
    }),
  );
}
